### ISSUE TITLE
Set up AppVeyor

### ISSUE THREAD

**Issue by kytrinyx** [exercism/cli#459]
We have some windows-specific behavior that it would be nice to have CI coverage for.

**Comment by kytrinyx** [comment]
@robphoenix I have a branch with a working appveyor build... meaning that it runs. Also, it fails on all sorts of dirty path-related stuff.

Would you have a chance to look at some of the failing tests on your windows machine and see if you can figure out what's going on?

The branch is `nextercism-appveyor`.

**Comment by kytrinyx** [comment]
I've messed around a bit more, and it looks like one of the biggest problems is that `filepath.Clean` doesn't normalize the volume name. When we do something like this:

```
_, cwd, _, _ := runtime.Caller(0)
```

Then `cwd` comes back with a lowercase volume name. But when we walk a directory, the results come back with an uppercase volume name.

**Comment by kytrinyx** [comment]
I take that back. When I tried normalizing the volume name, it broke in other interesting ways.

**Comment by kytrinyx** [comment]
Interestingly, it turns out this is directly related to not one but two open issues in the v2 feedback repo:

- https://github.com/exercism/v2-feedback/issues/42
- https://github.com/exercism/v2-feedback/issues/44

Right now I'm flailing around in the dark trying to figure out what the path-related best practices are.

Am I screwing this up because I'm comparing string-equality rather than using the filepath.Match() (which I just discovered a few hours ago)? Should we be using `path` rather than `filepath` in some places? I just don't know enough about this.

**Comment by robphoenix** [comment]
> @robphoenix I have a branch with a working appveyor build... 

Gah, sorry I didn't see this notification until now. I'll dig into this.

**Comment by kytrinyx** [comment]
> I'll dig into this.

Wonderful! Thank you ✨

**Comment by robphoenix** [comment]
If I open PR's to the appveyor branch, we'll be able to rebase out the *playing about* commits when it's ready to merge into `nextercism` right?

**Comment by kytrinyx** [comment]
Yepp, absolutely.

**Comment by robphoenix** [comment]
@kytrinyx Sorry, I've not forgotten about this, work/life/go-track blah, blah, blah. Just to let you know it's still on my radar.

**Comment by kytrinyx** [comment]
I know all sorts of things about the problems with work/life/exercism balance! Thanks for the ping :-)

**Comment by robphoenix** [comment]
> I know all sorts of things about the problems with work/life/exercism balance! 

Honestly, I don't know how you do it! Mad props.

**Comment by robphoenix** [comment]
Ok, tentatively, we have a passing Appveyor build. There are caveats to this.

Let me go through the changes.

1. https://github.com/exercism/cli/commit/f5eb086a21fc9f5db4cb09abd24a99b500cd6e51 - This is just plain cheating. I got these tests passing on my machine by creating a separate symlinked Windows directory. Unfortunately when pushed, this symlinked dir becomes a directory and so AppVeyor recognises it as a normal file. I'm pretty sure I tried creating the directory on the fly, but it was a few days ago and I can't be sure. I will spend some more time finding a solution to this
1. https://github.com/exercism/cli/commit/0601098d9f3700b6e45afbb6c6faa09c64ce945b - When the file is hidden on Windows we get an `Access Denied` error when we try to write to it. So we need to temporarily unhide it to write to it.
1. https://github.com/exercism/cli/commit/772771226091c2078af7b82eb8bb21afe1759e53 - I'm really not sure about this one. It gets the tests passing but I'm concerned it's just avoiding the actual problem. Interestingly, these tests pass on my machine without this change, but on the Appveyor build they fail because one set of paths come back with the volume as `C:\` and the other come back as `c:\`, where otherwise they are the same. I'm not certain that this hacky fix is hiding other places where this difference could trip up Windows users.

I haven't spent the time yet to check https://github.com/exercism/v2-feedback/issues/44, https://github.com/exercism/v2-feedback/issues/43 & https://github.com/exercism/v2-feedback/issues/42 against these changes, which may shed some light on how robust this passing build is. I should be able to get to that next week.

**Comment by kytrinyx** [comment]
>  There are caveats to this.

😂 Indeed! This is a great start, though. Thanks for tackling it, and for looking into the v2 issues.

**Comment by robphoenix** [comment]
update.

So some of the tests in `workspace_test.go` were failing because paths with a lowercase volume name (`c:\Users\`) were coming back from the `Locate` method with an uppercase volume name (`C:\Users`). 
This is actually a deliberate effect of `filepath.EvalSymlinks`.
See:
https://github.com/golang/go/blob/master/src/path/filepath/symlink_windows.go#L15
https://github.com/golang/go/issues/13980

So, to mitigate this I've wrapped Go's `EvalSymlinks` method in a function that de-normalises the path if necessary. I really don't know if this is the right approach or whether this is an awful hack or what, so any thoughts on this would be much appreciated!

**Comment by robphoenix** [comment]
I also appear to have somehow attached myself to your commits on the 7th of September. I have no idea how I did this. Sorry. I'll try to get to the bottom of it but if you know how to undo it I'd love to hear! :rofl: :roll_eyes:

**Comment by kytrinyx** [comment]
> So some of the tests in workspace_test.go were failing because paths with a lowercase volume name (c:\Users\) were coming back from the Locate method with an uppercase volume name (C:\Users).

Yes! I ran into that and was struggling to figure out what the right fix is. In my heart of hearts I keep wishing the Go project would normalize windows volume names consistently so that I don't have to understand the nuances of it. Selfish of me, I know. I just can't seem to figure out when one is more desirable or appropriate than the other!

> I also appear to have somehow attached myself to your commits on the 7th of September.

Heh. You're totally welcome to attach yourself to my commits—it simply means that I committed and you pushed (so presumably there was a force push involved somewhere). It always seems super friendly to me, like virtual async pairing.
